Added Fourier transform module

cczhu opened the pull request:
Added a Fourier transform wrapper to use different packages with a common interface.  Created wrappers for NumPy and pyFFTW.

Notes:

- Not sure if I've set up the `pyfftw` testing correctly or not...
- Having the user specify the frequency-domain dtype right now doesn't do anything for `NumpyFFT`, since we just use whatever pops out of `numpy.fft.fft`, etc.  We could typecast it, but that would add to the computational overhead.
- Not sure if `@_check_fft_exists` is even useful, since without it we still get something like 

```
AttributeError: 'NumpyFFT' object has no attribute '_fft'
```

- Should we support situations where we pass the frequency-domain dummy array and time-domain dtype (the inverse of what we currently have, to support dechannelization)?
- Not sure if we need Hermitian FFTs.

mhvk commented on .travis.yml:
```diff
@@ -57,6 +60,8 @@ before_install:
     # - export DISPLAY=:99.0
     # - sh -e /etc/init.d/xvfb start
+    - sudo apt-get update
+    - sudo apt-get install libfftw3-dev libfftw3-doc libfftw3-double3 libfftw3-long3 libfftw3-single3
```
Let's not bother with `long3`.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,290 @@
+    """
+
+    _data_format = {'time_shape': None,
+                    'time_dtype': None,
```
Would seem one would never need this: just same type to same type (with real <-> complex obvious).

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,290 @@
+
+        Parameters
+        ----------
+        time_data : `~numpy.ndarray` or dict
```
I would start by being strict in what can be put in, and expand only if we find it is actually useful. Here, wouldn't `size` (or `n`) and `dtype` suffice?

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,290 @@
+            alternatively give a dict with 'shape' and 'dtype' entries.
+        freq_dtype : str
+            dtype of frequency-domain data.
+        axes : int, tuple, or None, optional
```
Similarly, I'd hesitate to do `None` here. Let's start explicit. Indeed, I think for our streams, we only need 1-dimensional FTs, so I would start with those.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,290 @@
+        """
+        return self._data_format
+
+    def fft(self, a):
```
In my "split" implementation, this might be `__call__`.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,290 @@
+        """Placeholder for inverse FFT."""
+        raise NotImplementedError()
+
+    def fftfreq(self, a_length, sample_rate=None, positive_freqs_only=False):
```
I think `sample_rate` should come from `__init__` (fine to allow it to be `1.`) - similarly, why `a_length` when we know the size? I don't think this would need any arguments.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,290 @@
+    _axes = None
+    _norm = None
+
+    def setup(self, time_data, freq_dtype, axes=None, norm=None,
```
Even after looking through the rest, I don't understand why this isn't just `__init__`...

cczhu commented:
Based on discussion today: will switch from using `setup` to having dedicated factory classes for each FFT wrapper, with a single abstract `make_fft` function deciding which factory class to use for easy handling.  For example, a dedisperser will call
```
dd = Dedisperse(data, ..., make_fft('pyfftw', **kwargs))
```
`make_fft` will call `fft_maker = pyfftw_fft_maker(**kwargs)`, which returns a factory class instance.  `dd` is then able to generate FFTs by calling `fft = fft_maker(...)`, which returns a single forward or backward transform and its metadata.  Metadata is stored both for the user's sake and so that calling `ifft = fft.inverse()` is possible.

Also decided to drop multi-axis FFT support.

mhvk commented:
@cczhu - plan sounds good except that I think the factory maker should get as one as its arguments whether a forward or inverse transform is wanted (i.e., return one or the other, not both).

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+        output_shape = list(input_data['shape'])
+        # The inverse of a real-valued transform should return real values.
+        if direction == 'inverse':
+            assert 'complex' in input_data['dtype'].name, (
```
`dtype.kind == 'c'`

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+                "frequency-domain array must be complex.")
+            if real_transform:
+                output_shape[axis] = output_shape[axis] * 2
+                output_dtype = np.dtype('float{0:d}'.format(
```
Use `'f{0:d}` to avoid the factor 8.

cczhu commented:
Redesigned the Fourier module, though currently without a working test suite and with commented out code, etc.

Some design questions:

- I'm not sold on my naming scheme.  In particular, I'm not a fan of `_engine_name` as a string handle for each FFT maker class.
- I'm not sure about passing a whole bunch of variables around to make instances of `FFT`.  It might make more sense to pass a dict of metadata, which can then be deployed as kwargs to eg. `inverse`.
- If I create an instance of an FFT maker, eg. `maker = NumpyFFTMaker(...)`, doing `maker?` in IPython returns both the class and `__call__` docstrings, which is ugly.  Is there any way for only the call docstring to be returned?
- I would much rather have `get_fft` in `__init__.py`, even if it violates some design considerations, because it's too simple and too short of a function (even with an expanded docstring).

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+    def norm(self):
+        """Normalization convention.
+
+        As in `numpy.fft`, `None` is an unscaled forward transform and 1 / n
```
Here, I think it should just be defined (i.e., translate if `self._norm is None`)

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+        Uses `numpy.fft.fftfreq` for complex time-domain data, which returns,
+        for an array of length n and a time-domain ``sample_rate``,
+
+            f = [0, 1, ...,   n/2-1,     -n/2, ..., -1] * sample_rate
```
This is not true:
```
In [42]: np.fft.fftfreq(8)
Out[42]: array([ 0.   ,  0.125,  0.25 ,  0.375, -0.5  , -0.375, -0.25 , -0.125])
```

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+            sample_rate = 1.
+        a_length = self.data_format['time_shape'][self.axis]
+        if 'float' in self.data_format['time_dtype'].name:
+            return np.fft.rfftfreq(a_length) * sample_rate
```
Just put it `sample_rate`! I.e.,
```
np.fft.rfftfreq(a_length, 1./sample_rate)
```
this even works with units.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+
+                def fft(a):
+                    return np.fft.fft(a, axis=axis, norm=norm).astype(
+                        data_format['freq_dtype'])
```
Here and below, add `copy=False` in the `astype` (so that nothing gets done if the type is already OK).

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+from astropy.utils import lazyproperty
+
+
+class FFTMakerBase(object):
```
If we give this class a metaclass that tracks subclasses (or use the new `__init_subclass__` - maybe it is fine to requite python >=3.6...), then there is no need to keep the hand-crafted list in `__init__.py`.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+class FFTMakerBase(object):
+    """Base class for all FFT factory classes."""
+
+    _engine_name = None
```
Agreed this is ugly. Look at astropys `astropy.coordinates.representation` for another idea (see the `get_name` classmethod - basically strip any `FFTMaker` and lowercase the rest).

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+            a tuple.  If `None` (default), all axes are used.  For real-valued
+            time-domain data, the real-input transform is performed on
+            ``axes[-1]``.
+        norm : 'ortho' or None, optional
```
Maybe just have `ortho=True/False`?

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+                       'freq_shape': freq_data['shape'],
+                       'freq_dtype': freq_data['dtype']}
+
+        return self._setup_transform(data_format, direction, axis, norm,
```
I think the logic is slightly reversed here. I would suggest to move the code here to one or more methods that interpret input (possibly, these are "creation" methods, i.e., `from_input_array`, etc.), and make `__call__` give a `NotImplementedError`. That way subclasses override `__call__` which seems more logical.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+    def _setup_transform(data_format, direction, axis, norm, sample_rate):
+        raise NotImplementedError()
+
+    def inverse(self, data_format, direction, axis, norm, sample_rate):
```
I don't understand the point of this; or, at least not unless it would just take a FFT instance as its sole argument.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,273 @@
+
+    @property
+    def direction(self):
+        """Axis over which to perform the FFT."""
```
Copy & paste error in docstring.

mhvk commented:
@cczhu - I've thought a bit more about the structure, and think we can do better if the creation of the `fft` instance is a two-step process:
1. Create a new *subclass* of `FFT` and something numpy, pyfftw specific, which class attributes that describe the data shape and type. (One creates new classes using `type`; it's not very difficult.)
2. Have that class's `__init__` create an *instance* that actually does the forward or inverse transform, real or complex (ie., signature might be `__init__(self, direction, real_transform)`; it would essentially have the stuff from your current `_setup_transform`). That way the instance can easily generate an inverse by doing, e.g., `self.__class__('inverse', self.real_transform)`.

In this scheme, I'm not 100% sure whether `FFTMaker.__call__` should return the subclass or the instance. I'd be tempted to let it be the former, with no interpretation of arguments, and instead also create methods like `forward` and `inverse` (or maybe `create_fft`, `*rfft`, `*ifft`, and `*irfft`?) which might then take time shape and dtype, and frequency shape and dtype arguments).

mhvk commented:
Yes, the purpose is to have the relevant arguments in one place, and thus avoid the `self._parent` call.

And, indeed, the only reference to the class would be from its instance.

cczhu commented:
Revised Fourier module design to use class factories (rather than filling out an abstract base class) and addressed PR comments.  Still no tests, but @mhvk, could you have a look to see if the design makes sense and is to your liking?

Notes:
  * Implemented a metaclass-based registry rather than using `__init_subclass__`, partly because I like being able to customize the registry and make it visible to end-users, and partly because it's more straightforward to document for those that want to hack something.  (Also I was lazy and went with what I know.)
  * Not completely happy with having FFTMaker subclasses' `__call__` methods repeat the same few lines of code to parse the data format and process keywords.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,424 @@
+                .format(s=self, fmt=self.data_format))
+
+
+class NumpyFFTMaker(FFTMakerBase):
```
Maybe put in its own file?

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,424 @@
+            else:
+                self._fft = self._inverse_fft
+
+        NumpyFFT = type("NumpyFFT", (FFTBase,), {
```
I know I recommended this, but now that I see it, I wonder why we don't just do
```
class NumpyFFT(FFTBase):
    _data_format = data_format
    ...
    if data_format['time_type'].kind == 'c':
        def __call__(self, ...):
```
That's definitely clearer.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,424 @@
+        data_format = self.get_data_format(time_data=time_data,
+                                           freq_data=freq_data, axis=axis)
+
+        NumpyFFT = self._setup_fft_class(data_format, axis, ortho, sample_rate)
```
It's not clear to me why we don't just define the class here.

cczhu commented:
Revised module design based on PR comments, and added a test suite.  While it doesn't really test the implementation, I had a few tests of the data format verifier already implemented, and so kept those in.

Still need to go through the docstrings and write a short tutorial.

cczhu commented:
Hmm, had to rig the .travis.yml to install pyfftw from conda-forge, as per pyFFTW/pyFFTW#40.  Not sure if what I did is the most efficient.

mhvk commented on .travis.yml:
```diff
@@ -37,11 +37,15 @@ matrix:
 
         # Check for sphinx doc build warnings.
-        - env: SETUP_CMD='build_sphinx -w'
+        - env: SETUP_CMD='build_sphinx -w' PIP_DEPENDENCIES='pyfftw'
```
I think you can just use `CONDA_DEPENDENCIES='pyfftw'` and then you don't need the extra stuff below

cczhu replied:
I specifically need to extract pyfftw from conda-forge, so I'm not sure if I can write something like `CONDA_DEPENDENCIES='-c conda-forge pyfftw'`.

mhvk replied:
My point was that `conda` may just have packaged `pyfftw` themselves, and that that version will work. Worth trying...

mhvk replied:
Well, that clearly didn't work! So, I guess better remove that change...

mhvk commented on scintillometry/fourier/pyfftw.py:
```diff
@@ -0,0 +1,156 @@
+
+
+# Only instantiate the class if FFTW exists.
+try:
```
The `try/except` is not needed here, since you already do it in `fourier/__init__.py` (and it is OK to fail when someone attempts to import this directly).

cczhu commented on docs/fourier/index.rst:
```diff
@@ -0,0 +1,103 @@
+extracts the data type and shape (the dummy array is not used for the
+transform). We can also pass this data directly as a dictionary (see below). 
+We have chosen orthogonal normalization, which normalizes both the
+frequency and time-domain outputs by :math:`1 / n`, where :math:`n` is the
```
Yup, lost the {1/2} in translation.

cczhu replied:
Yup, I forgot to write the {1/2} (the docstring is correct, but the docs aren't).  Thanks for catching that.

cczhu commented on docs/fourier/index.rst:
```diff
@@ -0,0 +1,103 @@
+`~scintillometry.fourier.base.FFTMakerBase.ifft` to produce forward and inverse
+transform instances, respectively::
+
+    >>> fft2 = FFTMaker.fft(time_data={'shape': (1000,), 'dtype': 'float64'},
```
I've always agreed with this :)

I do think the class factory design (along with choosing directions at `__init__`) works well for creating inverse transforms, so I think your proposed solution is best.

cczhu commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,345 @@
+            raise TypeError("at least one of time or frequency-domain arrays "
+                            "must be passed in.")
+
+        # Extract information if user passed in dummy arrays.
```
One current example in the docs is:

```
ifft2 = FFTMaker.ifft(time_data={'shape': (1000,),
                                                   'dtype': 'float64'},
                                 ortho=True, sample_rate=1.*u.kHz)
```

so I can see the argument for not having a `freq_data` at all.  I suppose I was coding for the offhand case where one wants a frequency-to-time transform and doesn't want to be bothered figuring out what the time data should look like.  Happy to talk about this in pereson.

cczhu commented on docs/fourier/index.rst:
```diff
@@ -0,0 +1,103 @@
+create an inverse transform that outputs arrays with floating point dtypes,
+explicitly pass in time-domain array info::
+
+    >>> ifft2 = FFTMaker.ifft(time_data={'shape': (1000,),
```
Related to the broader discussion below, but I figured that if I had the machinery, I might as well allow for odd `n` as well.

cczhu commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,345 @@
+            direction=('forward' if self.direction == 'inverse'
+                       else 'inverse'))
+
+    def copy(self):
```
I wrote that after instinctively trying to use `fft.copy()` when playing with the module, so I figured why not implement it.

cczhu commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,345 @@
+    def __eq__(self, other):
+        # Assumes that class names are unique, which should be the case unless
+        # the user improperly initializes the class factory.
+        return (self.__class__.__name__ == other.__class__.__name__ and
```
We do (eg. someone hoping to farm off expensive FFTs to GPUs or something), but not when using `==`.  Indeed, it might be advantageous for ``fft1 == fft2`` to return `True` even if the backend is different.  Removed the first check.

cczhu commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,345 @@
+        return (self.__class__.__name__ == other.__class__.__name__ and
+                self.direction == other.direction and
+                self.data_format == other.data_format and
+                self.axis == other.axis and self.ortho == other.ortho and
```
Changed.

cczhu commented on scintillometry/fourier/base.py:
```diff
@@ -0,0 +1,345 @@
+        freq_shape[axis] = freq_shape[axis] // 2 + 1
+        return tuple(freq_shape)
+
+    def __call__(self, **kwargs):
```
Agreed, and changed.

cczhu commented on scintillometry/fourier/numpy.py:
```diff
@@ -0,0 +1,103 @@
+                                           freq_data=freq_data, axis=axis)
+
+        # Prepare either fft or rfft functions.
+        if data_format['time_dtype'].kind == 'c':
```
Funnily enough I kept it outside to make it similar to the old pyfftwFFTMaker, but then changed that so `__init__` handles everything.  Changed.

cczhu commented on scintillometry/fourier/numpy.py:
```diff
@@ -0,0 +1,103 @@
+            _norm = 'ortho' if ortho else None
+            _sample_rate = sample_rate
+
+            _forward_fft = forward_fft
```
We also lose some complexity, since we know both the direction and whether the FFT is real, so only need `_fft`.

cczhu commented:
As discussed today, we're switching out the dict of time and frequency array information for two variables storing the shape and dtype of the array in time, respectively.  This is enough to constrain both the forward and backward transforms.  We're also dropping the `fft` and `ifft` convenience methods, and making `__call__` always return an instance rather than a class definition (the user will now have to also pass the direction of the transform).

cczhu commented:
Addressing PR comments.  Most notably, changed the `FFTMaker` API so that only the time-domain array shape and dtype are passed (and must be passed), and altered `__call__` to return an instance of an FFT class rather than the class definition.  Also:

* Eliminated the `data_format` dictionary in favour of four private variables.  Kept the frequency-domain dtype and shape because it's a pain to calculate these again when the inverse is requested.
* Greatly shortened the FFT class docstrings - since the user shouldn't directly be accessing the class `__init__`, it only serves to clutter up the docs when `?` is used.  Also made some general improvements to the docstrings and docs.

cczhu commented on docs/fourier/index.rst:
```diff
@@ -0,0 +1,86 @@
+    <Quantity [0.   , 0.001, 0.002, 0.003, 0.004, 0.005, 0.006, 0.007, 0.008,
+           0.009] kHz>
+
+For multi-dimensional arrays, the sample frequencies are for the transformed
```
As in, if you RFFT'ed along the first axis of something that's (100, 25, 25), `freq` should have shape (50, 25, 25), with `freq[i, :, :]` all having the same value?

mhvk replied:
No, shape `(50, 1, 1)` - we'd of course rely on numpy broadcasting to do the rest! (But people could then add a base frequency that did depend on channel/polarization/whatever)

cczhu commented on scintillometry/fourier/pyfftw.py:
```diff
@@ -0,0 +1,141 @@
+            _fftw_kwargs = self._fftw_kwargs
+
+            def __init__(self, direction='forward'):
+                super(PyfftwFFT, self).__init__(direction=direction)
```
I think I was afraid of scope issues (that it would call `super(PyfftwFFTMaker, self)` instead).  Changing.

cczhu commented on scintillometry/fourier/tests/test_fourier.py:
```diff
@@ -0,0 +1,91 @@
+
+    def setup(self):
+        """Pre-calculate NumPy FFTs."""
+        x = np.linspace(0., 10., 7919)    # 7919 is the 1000th prime number!
```
It was more that 7919 was odd, which would check that our real transforms are self-consistent with `numpy.fft.rfft` ones.  I just picked a prime for fun.  (I can remove that comment if it's confusing.)

mhvk replied:
No, it's OK, just let it be.

cczhu commented:
Okay, added dimension expansion for frequencies, and ready for final review before merging.

mhvk commented on scintillometry/fourier/base.py:
```diff
@@ -151,13 +151,20 @@ def freq(self):
+        else:
+            freqs = np.fft.fftfreq(a_length, d=(1. / sample_rate))
+        # Add trailing dimensions.
+        freqs_shape = (freqs.shape +
```
In this case, it may be quite reasonable to do the change in-place, i.e.,
```
freqs.shape = (freqs.shape + ....
```

cczhu replied:
You can do that!?  Changing...

mhvk commented on scintillometry/fourier/tests/test_fourier.py:
```diff
@@ -53,8 +56,8 @@ def test_fft(self, key):
         # Check frequency.
-        assert_quantity_allclose(fft.freq, np.fft.fftfreq(
-            len(self.y_exp), d=(1. / self.sample_rate)))
+        assert_quantity_allclose(fft.freq.squeeze(), self.freq_Y_exp)
```
No need for the squeeze is this is the expected shape.

cczhu replied:
Isn't `self.freq_Y_exp` a different shape though?  Or should I expand those in the setup instead of also checking the shape below line 59?

cczhu replied:
Never mind, just saw below.

mhvk commented on scintillometry/fourier/tests/test_fourier.py:
```diff
@@ -88,4 +91,5 @@ def test_fft(self, key):
 
         # Check frequency.
-        assert_quantity_allclose(fft.freq, np.fft.fftfreq(len(self.y_3D[1])))
+        assert_quantity_allclose(fft.freq.squeeze(), self.freq_Y_3D)
```
Can we just assert `fft.freq, self.freq_Y_3D[:, np.newaxis]`?

cczhu commented:
Could have sworn I already updated the PR.  Oh well, now I've uploaded changes.
